File: backend/src/scoring/love_languages.rs
use std::collections::HashMap;

use crate::data::load_test_data;
use crate::error::{AppError, AppResult};
use crate::interpretation;

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::load_test_data;

    fn make_answers(value: f64) -> HashMap<u32, f64> {
        let data = load_test_data("love_languages").unwrap();
        let questions = data["questions"].as_array().unwrap();
        let mut answers = HashMap::new();
        for q in questions {
            let qid = q["id"].as_u64().unwrap() as u32;
            answers.insert(qid, value);
        }
        answers
    }

    #[test]
    fn test_love_languages_has_required_fields() {
        let answers = make_answers(4.0);
        let result = score(&answers, "en").unwrap();
        let obj = result.as_object().unwrap();
        assert!(obj.contains_key("scores"));
        assert!(obj.contains_key("primary_language"));
        assert!(obj.contains_key("secondary_language"));
        assert!(obj.contains_key("language_ranking"));
        assert!(obj.contains_key("profile_summary"));
    }

    #[test]
    fn test_love_languages_scores_in_range() {
        let answers = make_answers(3.0);
        let result = score(&answers, "en").unwrap();
        let scores = result["scores"].as_object().unwrap();
        for (lang, val) in scores {
            let v = val.as_f64().unwrap();
            assert!(v >= 0.0 && v <= 100.0, "language {} score {} out of [0,100]", lang, v);
        }
    }

    #[test]
    fn test_love_languages_primary_high() {
        let answers = make_answers(5.0);
        let result = score(&answers, "en").unwrap();
        let primary = result["primary_language"].as_str().unwrap();
        let scores = result["scores"].as_object().unwrap();
        let primary_score = scores.get(primary).and_then(|v| v.as_f64()).unwrap();
        // With all 5s, primary should be one of the languages
        assert!(primary_score > 0.0);
        assert!(result["secondary_language"].as_str().is_some());
    }

    #[test]
    fn test_love_languages_spanish() {
        let answers = make_answers(2.0);
        let result = score(&answers, "es").unwrap();
        assert!(result["profile_summary"].as_str().unwrap_or("").len() > 0);
    }

    #[test]
    fn test_love_languages_all_languages_present() {
        let answers = make_answers(3.0);
        let result = score(&answers, "en").unwrap();
        let scores = result["scores"].as_object().unwrap();
        for lang in &["words_of_affirmation", "quality_time", "receiving_gifts", "acts_of_service", "physical_touch"] {
            assert!(scores.contains_key(*lang), "missing language: {}", lang);
        }
    }
}

fn normalize(value: f64, min: f64, max: f64) -> f64 {
    ((value - min) / (max - min)) * 100.0
}

const DIMENSIONS: &[&str] = &[
    "words_of_affirmation",
    "quality_time",
    "receiving_gifts",
    "acts_of_service",
    "physical_touch",
];

pub fn score(answers: &HashMap<u32, f64>, language: &str) -> AppResult<serde_json::Value> {
    let data = load_test_data("love_languages").map_err(|e| {
        AppError::Internal(format!("Failed to load love_languages data: {}", e.0))
    })?;
    let questions = data["questions"].as_array().ok_or_else(|| {
        AppError::Internal("love_languages data missing questions array".to_string())
    })?;

    let mut dim_raw: HashMap<&str, Vec<f64>> = HashMap::new();
    for d in DIMENSIONS {
        dim_raw.insert(d, Vec::new());
    }

    for q in questions {
        let qid = q["id"].as_u64().unwrap_or(0) as u32;
        if let Some(&value) = answers.get(&qid) {
            let dim = q["dimension"].as_str().unwrap_or("");
            dim_raw.entry(dim).or_default().push(value);
        }
    }

    let mut scores_map = serde_json::Map::new();
    for d in DIMENSIONS {
        let values = dim_raw.get(d).cloned().unwrap_or_default();
        let score = if values.is_empty() {
            0.0
        } else {
            let mean = values.iter().sum::<f64>() / values.len() as f64;
            (normalize(mean, 1.0, 5.0) * 10.0).round() / 10.0
        };
        scores_map.insert(d.to_string(), serde_json::json!(score));
    }

    // Sort by score descending to get ranking
    let mut ranked: Vec<(&str, f64)> = DIMENSIONS.iter()
        .filter_map(|d| {
            let score = scores_map.get(*d).and_then(|v| v.as_f64())?;
            Some((*d, score))
        })
        .collect();
    ranked.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

    let primary = ranked.first().map(|(k, _)| *k).unwrap_or("words_of_affirmation");
    let secondary = ranked.get(1).map(|(k, _)| *k).unwrap_or("quality_time");

    let primary_score = scores_map.get(primary).and_then(|v| v.as_f64()).unwrap_or(0.0);
    let secondary_score = scores_map.get(secondary).and_then(|v| v.as_f64()).unwrap_or(0.0);

    let summary = if language == "en" {
        format!(
            "Primary Love Language: {} ({:.0}%). Secondary: {} ({:.0}%).",
            primary.replace('_', " "),
            primary_score,
            secondary.replace('_', " "),
            secondary_score,
        )
    } else {
        let lang_names = data["dimensions"].as_object().unwrap();
        let primary_name = lang_names.get(primary)
            .and_then(|v| v.get("es"))
            .and_then(|v| v.as_str())
            .unwrap_or(primary);
        let secondary_name = lang_names.get(secondary)
            .and_then(|v| v.get("es"))
            .and_then(|v| v.as_str())
            .unwrap_or(secondary);
        format!(
            "Lenguaje del Amor Principal: {} ({:.0}%). Secundario: {} ({:.0}%).",
            primary_name,
            primary_score,
            secondary_name,
            secondary_score,
        )
    };

    let scores_hash: HashMap<String, f64> = scores_map.iter()
        .filter_map(|(k, v)| v.as_f64().map(|f| (k.clone(), f)))
        .collect();

    let interp = interpretation::get_interpretation("love_languages", &scores_hash, language);
    let recs = interpretation::get_recommendations("love_languages", &scores_hash, language);

    let mut result = serde_json::Map::new();
    result.insert("scores".into(), serde_json::json!(scores_map));
    result.insert("primary_language".into(), serde_json::json!(primary));
    result.insert("secondary_language".into(), serde_json::json!(secondary));
    result.insert("language_ranking".into(), serde_json::json!(
        ranked.iter().map(|(k, v)| serde_json::json!({"language": k, "score": v})).collect::<Vec<_>>()
    ));
    result.insert("profile_summary".into(), serde_json::json!(summary));
    result.insert("interpretation".into(), interp);
    if let Some(obj) = recs.as_object() {
        for (k, v) in obj {
            result.insert(k.clone(), v.clone());
        }
    }
    Ok(serde_json::Value::Object(result))
}

File: backend/app/scoring/love_languages.py
import json
from pathlib import Path


_DATA_DIR = Path(__file__).resolve().parent.parent / "data"

with open(_DATA_DIR / "love_languages_questions.json", encoding="utf-8") as _f:
    _DATA = json.load(_f)

_DIMENSIONS = _DATA["dimensions"]
_QUESTIONS = _DATA["questions"]
_SCALE_MIN = _DATA["scale"]["min"]
_SCALE_MAX = _DATA["scale"]["max"]

_DIM_ITEMS: dict[str, list[int]] = {}
for q in _QUESTIONS:
    _DIM_ITEMS.setdefault(q["dimension"], []).append(q["id"])


def _normalize(raw_avg: float) -> float:
    return round(((raw_avg - _SCALE_MIN) / (_SCALE_MAX - _SCALE_MIN)) * 100, 1)


_DIM_LABELS = {
    "words_of_affirmation": {"es": "Palabras de Afirmación", "en": "Words of Affirmation"},
    "quality_time": {"es": "Tiempo de Calidad", "en": "Quality Time"},
    "receiving_gifts": {"es": "Recepción de Regalos", "en": "Receiving Gifts"},
    "acts_of_service": {"es": "Actos de Servicio", "en": "Acts of Service"},
    "physical_touch": {"es": "Contacto Físico", "en": "Physical Touch"},
}


def score_love_languages(responses: dict[int | str, int], lang: str = "es") -> dict:
    raw_scores: dict[str, list[float]] = {d: [] for d in _DIM_ITEMS}

    valid_ids = {q["id"] for q in _QUESTIONS}

    for qid_str, raw_val in responses.items():
        qid = int(qid_str)
        if qid not in valid_ids:
            continue
        for dim, items in _DIM_ITEMS.items():
            if qid in items:
                clamped = max(_SCALE_MIN, min(_SCALE_MAX, int(raw_val)))
                raw_scores[dim].append(float(clamped))
                break

    scores: dict[str, float] = {}
    for dim in _DIM_ITEMS:
        vals = raw_scores[dim]
        if vals:
            avg = sum(vals) / len(vals)
        else:
            avg = _SCALE_MIN
        scores[dim] = _normalize(avg)

    labeled_scores = {}
    for dim, key in _DIM_LABELS.items():
        labeled_scores[key[lang if lang in key else "en"]] = scores[dim]

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)
    primary_key = ranked[0][0]
    secondary_key = ranked[1][0]

    primary_lang = _DIM_LABELS[primary_key][lang if lang in _DIM_LABELS[primary_key] else "en"]
    secondary_lang = _DIM_LABELS[secondary_key][lang if lang in _DIM_LABELS[secondary_key] else "en"]

    language_ranking = [
        {
            "language": _DIM_LABELS[dim][lang if lang in _DIM_LABELS[dim] else "en"],
            "score": score,
        }
        for dim, score in ranked
    ]

    return {
        "scores": labeled_scores,
        "primary_language": primary_lang,
        "secondary_language": secondary_lang,
        "language_ranking": language_ranking,
    }
